### DISCUSSION TITLE
Uncaught ReferenceError: Alpine is not defined / How to use with CDN??

### DISCUSSION THREAD

**Discussion by Sjoerd82** [alpinejs/alpine#2828] [1. Help] [answered]
Could someone explain to me how to use the Alpine in a CDN-context?

I get the error that "Alpine" is not defined yet". Somewhere here on this issues board (can't find it anymore) this is discussed, and it seems to be a initialization/timing issue that can be resolved by waiting for the init-event. However, this I think, limits the scope if you nest the data declarations in that event handler.

I plan to use this in conjunction with EventSource-handlers, so I'd like to use my variables anywhere within my script.

		<script defer src="https://unpkg.com/alpinejs@3.x.x/dist/cdn.min.js"></script>
		<script type="text/javascript">
			console.log('Load data script');

			//document.addEventListener('alpine:init', () => {
				console.log('alpine:init');
				Alpine.data('example', () => ({
					a: 'b',
					init() {
						console.log('Example loaded');
					}
				}))
			//});

		        // I want to use my data here

		</script>

I know that the CDN-path is not really the way to go, normally I use bundlers, but in this particular scenario I cannot use it, so am going with CDN. I feel that using Alpine this way is overly difficult (and thusfar unpleasant, due to lack of documentation). If there is a simple way to use Alpine via CDN I'd love to hear it!

**Comment by SimoTod** [comment, marked as answer]
If the script is deferred it runs after the full page has been parsed.
In your example, Alpine is only available after the 'alpine:init' event which is a way to tell the browser that the script has been loaded.

If you remove defer, it will become a blocking script then you can access Alpine without wrapping the code in the event listener but you need to move those script tags at the bottom of the body.

**Reply by abdellahkaba** [reply]
Hello everyone I have a problem I hope you could help me the error is that formData is not define

 <form action="/contact/submit" method="POST" 
                        x-data="{formData: {name: '',email: ''}}" x-on:submit.prevent="submitForm">
                        
     </form>

**Reply by ekwoka** [reply]
Make your own discussion if it's not related, and nothing in your code suggests you even tried to access `formData`

**Comment by Sjoerd82** [comment]
Yes, removing defer to make the script blocking works. I had to put any HTML that uses the data after it, otherwise the referenced variables were also undefined. So effectively put it on top of the body.

However, follow-up question, as I am starting to realize the data scope is still limited, I think.. I cannot seem to access it from within another eventhandler.

The entiry page looks like this. There's an event handler that fires every time a message arrives.
As a proof-of-concept I am trying to increase the counter each time. This should probably be a store rather than a data object, but that might be moot, as I cannot seem to access the Alpine data at all. I hope I'm doing something wrong:

```
<html>
	<body>
		
		<script src="https://unpkg.com/alpinejs@3.x.x/dist/cdn.min.js"></script>
		<script type="text/javascript">
			console.log('alpine starts');
			Alpine.data('example', () => ({
				counter: 0,
				init() {
					console.log('Example loaded');
				}
			}))				
		
			let evtSse; // EventSource
			
			//Check for Browser feature compatibiliity
			if (!!window.EventSource) {
				evtSse = new EventSource('listen');  //instantiate the Event source
			} else {
				alert("You're browser does not support EventSource");
			}
			
			//Now bind various Events , Message, and Error to this event
			evtSse.addEventListener('open', function(e) {
				console.log("Connection was opened.")
			}, false);
			
			evtSse.addEventListener('message', function(e) {
			  console.log(e.data);
			  
			  // I think our Alpine data may not be reachable from within the eventhanlder context?
			  console.log('got a msg, increase counter',Alpine)
			  //Alpine.data.counter = data.counter + 1 // "data is not defined"
			  //Alpine.$data.counter = data.counter + 1 // "data is not defined"
			  
			}, false);

			evtSse.addEventListener('error', function(e) {
				if (e.readyState == EventSource.CLOSED) { 
					console.log("Connection was closed. ");
				}
			}, false);
		</script>
		
		<div x-data="example">
			Counter: <span x-text="counter"></span>
		</div>
	</body>
</html>
```

**Reply by SimoTod** [reply]
I don't think you can use Alpine.data that way.  Alpine data is just a factory for your components. You should use a store.
Anyway, the data that is not defined is the one on the right side if the expression: Alpine.data.counter = **data**.counter + 1
Putting the script at the top is a recipe for disaster. It works with this page but you'll get tons of errors when you work on a real webpage because of the mutation observer trying to evaluate stuff when they are not ready yet.

Is there a reason why you can't wrap your full app into `document.addEventListener('alpine:init', () => {` (including the EventSource bit)? It looks like you are just using the tool in the wrong way.

**Comment by Sjoerd82** [comment]
And by "a store" you mean that Alpine.js' store can be used? (is it similar to VueX/Pinia?)? Or are you saying that to accomplish this I should add a separate state manager that is not Alpine?

I've converted "data" to "store", but still can't access it in the message event hander, also not when I wrap that event handler in the Alpine init callback as suggested.

If I put the HTML above the script-tag, I get _not-defined_ errors. I guess I could perhaps add an isLoaded-indicator variable, and set it to true once the init has been completed, I wouldn't expected to have to add this kind of control to make Alpine work (or is that because of using a CDN?).

```
<html>
	<body>

		<!-- cdn.min.js:1 Alpine Expression Error: example is not defined --
		<div x-data="example">
			Counter: <span x-text="counter"></span>
		</div>
		-->

		<script src="https://unpkg.com/alpinejs@3.x.x/dist/cdn.min.js"></script>
		<script type="text/javascript">
			console.log('alpine starts');
			//Alpine.data('example', () => ({
			Alpine.store('example', {
				counter: 0,
				init() {
					console.log('Example loaded');
					
					
					
					let evtSse; // EventSource
					
					//Check for Browser feature compatibiliity
					if (!!window.EventSource) {
						evtSse = new EventSource('listen');  //instantiate the Event source
					} else {
						alert("You're browser does not support EventSource");
					}
					
					//Now bind various Events , Message, and Error to this event
					evtSse.addEventListener('open', function(e) {
						console.log("Connection was opened.")
					}, false);
					
					evtSse.addEventListener('message', function(e) {
					  console.log(e.data);
					  
					  // I think our Alpine data may not be reachable from within the eventhanlder context?
					  console.log('got a msg, increase counter',Alpine)
					  //Alpine.store('counter') = Alpine.store('counter') + 1 // Invalid left-hand side in assignment, clearly we are not referencing the store here..
					  
					}, false);

					evtSse.addEventListener('error', function(e) {
						if (e.readyState == EventSource.CLOSED) { 
							console.log("Connection was closed. ");
						}
					}, false);
			
					
				}
			})

		</script>
		<div>
			Counter: <span x-text="$store.counter"></span>
		</div>
	</body>
</html>
```

**Reply by SimoTod** [reply]
it's because you are using it in the wrong way. Alpine.store('counter') doesn't exist. You have `Alpine.store('example').counter`.
This is the relevant piece of documentation: https://alpinejs.dev/globals/alpine-store#accessing%20stores

The Alpine.store itself should really be wrapped in the event listener and be defined before alpine kicks in otherwise `example` doesn't exist. (https://alpinejs.dev/globals/alpine-store#registering-a-store)

You can do it without the event listener and defer but you need to deeply understand how alpine works because it's not the canonical way to use it. In that case, you should spend some time on the source code.

**Comment by Sjoerd82** [comment]
Thanks for the hints, the code below now works. I'm by no means an expert, but the documentation could have some more examples (for example, the store accesssing format is not documented, or at least I can't find it. Which by the way doesn't take away from what seems to be a fine piece of software and a friendly community that is quick to answer. 👍 

Coming weeks I'll be expanding on this quick proof-of-concept, and will try to educate myself about deferring and timing of loading alpine.js and the rest of the DOM. For now, my thanks.

```
<html>
	<body>
		<script src="https://unpkg.com/alpinejs@3.x.x/dist/cdn.min.js"></script>
		<script type="text/javascript">
			console.log('alpine starts');
			Alpine.store('example', {
				counter: 0,
				init() {
					console.log('Example loaded');
					
					let evtSse; // EventSource
					
					//Check for Browser feature compatibiliity
					if (!!window.EventSource) {
						evtSse = new EventSource('listen');  //instantiate the Event source
					} else {
						alert("You're browser does not support EventSource");
					}
					
					//Now bind various Events , Message, and Error to this event
					evtSse.addEventListener('open', function(e) {
						console.log("Connection was opened.")
					}, false);
					
					evtSse.addEventListener('message', function(e) {
					  console.log(e.data);
					  Alpine.store('example').counter = Alpine.store('example').counter + 1
					  
					}, false);

					evtSse.addEventListener('error', function(e) {
						if (e.readyState == EventSource.CLOSED) { 
							console.log("Connection was closed. ");
						}
					}, false);
					
				}
			})
		</script>
		<div>
			Counter: <span x-text="$store.example.counter"></span>
		</div>
	</body>
</html>
```

**Reply by SimoTod** [reply]
No worries, the documentation is open source if you can see anything to improve. It's in the same repo inside packages/docs.
